### ISSUE TITLE
[LineChart]: support y-axis basepoint starting above zero for large values

### ISSUE THREAD

**Issue by scottdickerson** [carbon-design-system/carbon-charts#402]
#### I am submitting a...

- [x] Feature request
- [ ] Design defect
- [ ] Source code defect
- [ ] Demo/documentation defect
- [ ] Other

#### charts version:
0.16.25

### Issue description
For graphs that have large y-values that hover around the same range, it would be nice to have a broken y-axis (or a y-axis that doesn't start from 0) so that we can see more variance in the data

Here's an example:
![image](https://user-images.githubusercontent.com/6663002/68342615-9c9cfe80-00b0-11ea-9d3e-a3256b93ca25.png)

**Comment by theiliad** [comment]
Currently this is what happens:
![image](https://user-images.githubusercontent.com/14989804/69979147-b8eb5a00-14fb-11ea-9150-478a8f45b1ce.png)
(0 gets added to the scale domain)

we might need a `min` & `max` value in place in the options, defaulted to `0` and `null` respectively, that you can customize.

**Comment by anne-stevens** [comment]
I support this request.  It is good practice for line charts to NOT always start with zero on the y-axis.  As Scott said, if the y-axis data range better suited the range in data, then it would make the chart more legible and understandable.

**Comment by caesarsol** [comment]
In addition to #491 we'd like to add an option to include/exclude the `0`.
Our designers at Accurat (cc @gvnmagni) prefer to always show the 0 with a truncated axis, so in case of non-included 0 they would show the 0 with a truncation.

However, I would tackle the "Axes Truncation" problem in a separated issue, what do you think @theiliad?

**Comment by theiliad** [comment]
> In addition to #491 we'd like to add an option to include/exclude the `0`.
> Our designers at Accurat (cc @gvnmagni) prefer to always show the 0 with a truncated axis, so in case of non-included 0 they would show the 0 with a truncation.
> 
> However, I would tackle the "Axes Truncation" problem in a separated issue, what do you think @theiliad?

Agree with all of that Caesar

**Comment by ryshah** [comment]
+1 for allowing configuration option that allows axis to respect min/max ranges specified...

**Comment by gvnmagni** [comment]
Hi! Generally speaking, we support the idea of allowing the y-axis to start a different values according to data.

We'd like to get some feedback on how to visually deal with it, we have few ideas:

1. Usual axis, instead of starting at zero it would start at the desired valued (top-left option). The most clean and elegant option of all

Nonetheless it could be interesting to emphasize this behaviour, in order to always remind the user of this specific feature the axis could start at zero, make a jump and start again from the desired starting point. We have two options here:

2. Use the little wave to point out this feature, this would be consistent with the jumps design made previously (top-right)
3. Use double bars. This would be much clearer to the user since it is a much more common way of dealing with it and it would our choice (if we like the idea of always showing the zero value)

Any preference? @mjabbink

![Truncating Axis](https://user-images.githubusercontent.com/18484576/74670782-2aa57c80-51aa-11ea-9107-be08dcbc4740.png)

**Comment by mjabbink** [comment]
Little curves work better I think

—
Mike Abbink
Executive Creative Director,
IBM Brand Experience & Design
IBM Distinguished Designer

IBM

—
Sent via iPhone

**Comment by theiliad** [comment]
I'm assuming that we don't want to allow breaks within ranges where a graph is being shown:
![image](https://user-images.githubusercontent.com/14989804/74672376-257d0900-517b-11ea-970f-621a3a6fbfd5.png)

Seems a bit misleading to me

**Comment by gvnmagni** [comment]
We agree! We find it misleading as well

**Comment by theiliad** [comment]
We won't really be able to stop users from doing that (you could mess it up if you choose to do so)

I think this might be an item where we should also focus on providing the correct guidelines within our dataviz pages

**Comment by anne-stevens** [comment]
For line charts, starting the y-axis at a non-zero value is well accepted in data viz best practice.  I vote for Accurat's top left option b/c adding the zero value to the y-axis adds no value to the axis or viz.  

Bar charts are different.  Not starting the y-axis at zero is considered a visualization lie.  Having a gap in the data range on the y-axis that only applies to 2 bars would similarly be a lie.  IMO it would make IBM look bad to have products with bar charts such as the one in Eliad's comment above.  I don't think we should make that possible, even if it is covered in the guidance.

**Comment by theiliad** [comment]
> For line charts, starting the y-axis at a non-zero value is well accepted in data viz best practice. I vote for Accurat's top left option b/c adding the zero value to the y-axis adds no value to the axis or viz.
> 
> Bar charts are different. Not starting the y-axis at zero is considered a visualization lie. Having a gap in the data range on the y-axis that only applies to 2 bars would similarly be a lie. IMO it would make IBM look bad to have products with bar charts such as the one in Eliad's comment above. I don't think we should make that possible, even if it is covered in the guidance.

In my experience the guidance is the most effective tool we can use to highly discourage teams against producing such visualizations, however teams can still go against our recommended guidelines.

What you're talking about is definitely a concern.

**Comment by gvnmagni** [comment]
We agree with Anne and we suggest to do both:

1. remove the availability of this feature for barcharts
2. clearly point this out on the Carbon System pages  

To do this, the best thing would be to review/update the **Breaks in axes** paragraph, included in the **Axes and Labels** page, since its first example shown is the barchart itself, leaving only Linecharts examples

**Comment by mjabbink** [comment]
Agree with Anne’s comments

**Comment by theiliad** [comment]
While removing the functionality seems reasonable from a design perspective, it's not a great idea from a dev perspective.

We should avoid specific condition definitions within different modules in the code, these sort of additions will complicate the codebase and limit scalability.

I would at the very most suggest adding a console warning if the user tries to use this functionality on a bar chart, but really rely on the design teams to not allow this behaviour into their product.

@zvonimirfras @cal-smith @tw15egan

**Comment by tw15egan** [comment]
I agree with @theiliad and @cal-smith that a console warning/type error should be emitted if a user tries to use this functionality on a bar chart. This alongside proper documentation and guidelines (do's / don'ts) should prevent this behavior from the end-users

**Comment by zvonimirfras** [comment]
We can/should only do sensible defaults and possibly console warnings on the charts we ship.
Users can use our library to create custom charts which can do things we haven't foreseen. That's the power of the system we've built.

That's where the design guidelines come into place.

We can't (and shouldn't) police what people can build with custom charts using our library.
We can (and should) ship the sensible default, built-in, charts that follow those guidelines when used properly.

To Eliad's point, we can, in some cases, try warning them in case they are using them improperly but it will be counterproductive if we get into business of stopping them from doing things.

**Comment by anne-stevens** [comment]
Call me a purist, but I can't agree with the proposal that we build in the ability to interrupt the y-axis scale for a bar chart, but then build in a warning that says "Just kidding.  You probably don't want to do that."  I feel that we would do a better job of setting the end user up for success, out of the box, by (a) just not letting them interrupt the y-axis scale for a bar chart in the first place and (b) starting the y-axis scale at 0 by default for bar charts.  This is viz best practice.  

Otherwise we risk ending up on Visualization-WTF like this: https://viz.wtf/post/185041906238/not-as-cheap-as-youd-think#notes

**Comment by anne-stevens** [comment]
This convo started as a FR for NOT starting the y-axis at zero for a line chart.  I do support this, as it is standard practice in visualization.  Again, call me a purist, and with respect to contrary opinions about creating exceptions above.

**Comment by gvnmagni** [comment]
We have to say that, even if we perfectly understand the code logic behind the behaviour suggested for barcharts, we 100% agree with Anne. We are afraid that a warning, an error message or a paragraph within the guidelines will not be enough. In this specific case if users don't follow our suggestion they are going to make a clear mistake in representing data rather than a less functional chart

**Comment by caesarsol** [comment]
The frontend error is not very clean, and the console warning option – in my experience – gets ignored most of the times. 

I think we can change the options logics to make it impossible for barcharts to be configured in that way. It would be the best option both design-wise and developer-experience-wise.

It could make the code more complicated, but the library will eventually need chart-specific configuration options.

(I would not delay the PR to include this, however. It requires some work and planning with Eliad)

**Comment by theiliad** [comment]
@caesarsol yes let's not delay the `non-zero starting point` PR.

Axis breaks should be a separate PR anyways and hopefully down the road. We've got other requested functionality that seem to be a higher priority

**Comment by cal-smith** [comment]
To be clear, I'm not suggesting that we build in the capability to have breaks in a bar chart, and then surprise users by throwing warnings and errors all over the place. If the bar chart ends up with the ability to have axis breaks _as a side effect of adding them in general_ then we shouldn't increase complexity just to prevent users from doing that. If it's feasible to prevent users from enabling axis breaks in bar charts then awesome - lets do that. But we shouldn't compromise capability because some fractional percentage of users will misuse the capabilities of the library - they'll just do it a different way anyway.

**Comment by theiliad** [comment]
> To be clear, I'm not suggesting that we build in the capability to have breaks in a bar chart, and then surprise users by throwing warnings and errors all over the place. If the bar chart ends up with the ability to have axis breaks _as a side effect of adding them in general_ then we shouldn't increase complexity just to prevent users from doing that. If it's feasible to prevent users from enabling axis breaks in bar charts then awesome - lets do that. But we shouldn't compromise capability because some fractional percentage of users will misuse the capabilities of the library - they'll just do it a different way anyway.

Agreed. It'd basically require axes to know about graphs and know what graphs are being rendered which isn't ideal
